### PR TITLE
Fix #5473: Add wiki guide for upgrading the app's compile/target SDK versions

### LINKED ISSUE #5473: [Feature Request]: Provide instructions for updating the app's target API level and verifying compatibility.
### Describe the bug

We are getting regular emails from the Play Store roughly once a year that say:

"[Action required] Your app is affected by Google Play's target API level requirements"

These emails tell us that the app is targeting an old version of Android, and that we need to "meet target API level requirements".

The aim of this issue is threefold:

- [x] Write a wiki page that gives instructions on how to do this in the future (including the data-gathering process), so that any developer on the team can do it. (@BenHenning to do this.) 
- ~Fix the current app to correct the target API level requirements. Per the Google Play email, this must be done by 31 Aug 2024.~
- ~Establish an annual cadence for making these updates, such that they are completed well before we are warned about them by the Play Store.~

### Steps To Reproduce

N/A

### Expected Behavior

We should hopefully stop getting these emails if the app API level requirements are always correctly up-to-date.

### Screenshots/Videos

_No response_

### What device/emulator are you using?

N/A

### Which Android version is your device/emulator running?

N/A

### Which version of the Oppia Android app are you using?

N/A

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation

Fixes #5473

This PR introduces a new wiki page that describes the process by which we can carefully upgrade the app to a newer compile/target SDK version (with a specific emphasis on the target SDK version).

This is largely meant to capture past processes we have followed since those weren't documented in an easily reproducible way. The intent is for this now-documented process to be trialed by upgrading the app to SDK 34 in order to address #5535 (though that technically means that part (1) of the documented process has already been completed).

This PR does not:
- Address exactly when the compile version should be updated, except when required by the target SDK version. It's expected that the team will perform an out-of-band analysis for cases when the compile SDK actually needs to be upgraded sooner than the target SDK.
- Specify a strongly enforced cadence for upgrading the target SDK since the Google Play Console and Google Play app policy seems to already perform this function, though it does add a recommendation for checking for updates periodically (July/August every year).

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- This is introducing a new wiki page and doesn't change the app itself.

### CODE REVIEW

**Review by BenHenning — COMMENTED** [review]
Self-reviewed changes (brief skim).

**Comment by BenHenning** [comment]
PTAL @adhiamboperes for code owners.

PTAL @Rd4dev for just general thoughts on whether the wiki page is well-organized and well-written, and if you have any follow-up suggestions. Please feel free to consider this review optional. If you have the time, I'd appreciate any thoughts you have on this page.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: It looks like the illustration content might be getting cut off due to the text length. Could using a `<br>` tag help with this? (or other approach) 

```mermaid
flowchart TD
  E(Part 5: Submitting the upgrade) -->|Compile/target SDK PR submitted| F(Block)
```

with `<br>` tag:

```mermaid
flowchart TD
  E(Part 5: Submitting<br>the upgrade) -->|Compile/target SDK<br>PR submitted| F(Block)
```

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: >(similar to #5535)

Would it be preferable to include direct links than issue numbers, as MD files don’t seem to link them directly?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md:41]: Would splitting them into separate subheadings, for eg. "High-Level Considerations" and "Post Findings," improve structure and enhance readability? - [sample md](https://github.com/Rd4dev/Oppia-Android-Fork-from-Fork/blob/Rd4dev-patch-3/wiki/up.md#part-2-auditing-the-android-os-changelog)

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: Nit: Spelling of functionality
```suggestion
- Look for changes in permissions. This could either be an old permission that's now more restricted, or existing SDK functionality that's now blocked by a new permission (both have occurred in past Android OS updates). If the functionality and/or permission relates to Oppia Android, it should be noted. Note that some things may be tied to permissions and access control that's not obvious such as:
```

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: Would it be helpful to include somewhere a link to the [Bazel Build instructions wiki](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) here for building the local Bazel production app, or can we assume developers are already familiar with the process?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: Would it be beneficial to add subheadings such as 
- Deployment 
- User Behavior Testing 
- General Analysis
- Filing Issues

[sample md](https://github.com/Rd4dev/Oppia-Android-Fork-from-Fork/blob/Rd4dev-patch-3/wiki/up.md#part-3-testing-the-app--filing-problems)

or would it be preferable to maintain the current single flow?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: Same clarification as above, if adding subheadings for better structure would be helpful (may be)? 

- Address Identified Problems
- Verification of Fixes
- Final Confirmation

[sample md](https://github.com/Rd4dev/Oppia-Android-Fork-from-Fork/blob/Rd4dev-patch-3/wiki/up.md#part-4-fixing-and-stabilizing-support)

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: I'm not sure if this is needed or if I’m following the structure correctly, but even though Part 3 (Testing) refers to Part 5 (Upgrading), would it be useful to restructure and include a **'How to Upgrade'** section before the testing section (Part 3).

As I believe testing typically follows the upgrade process, providing the 'How to Upgrade' section earlier might create a clearer path. 

1. Identifying the need to upgrade
2. Auditing the Android OS changelog
3. **How to upgrade**
4. Testing the app & filing problems
5. Fixing and stabilizing support
6. Submitting the upgrade
7. Future work items and upgrading Robolectric

Or is the current structure designed to ensure that all necessary testing procedures are addressed before proceeding with the upgrade?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: Would the WORKSPACE file also require an update in this case? (apologies if I overlooked)

**Review by Rd4dev — COMMENTED** [review]
Thanks @BenHenning, would just like to understand two minor things:

1. Would it be useful to restructure/include a 'How to Upgrade' section?
2. Would the WORKSPACE file need an update?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: What exactly is being cut off for you?

Here's what I see:

image

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: I thought about doing that, and it's a good suggestion. Done everywhere on the page.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md:41]: Added new sections as suggested, though I went with 'tips for auditing' and kept the actual high-level instructions in the top-level section.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Nice catch! Fixed.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Linking to the Bazel documentation seems reasonable--done.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Good question--it shouldn't anymore now that we've moved relevant build configurations to build_vars.bzl. It does still need to be updated for the target SDK, but that'll go away soonish (once we remove the APKs targets--they are rarely used and we should prefer using the AAB builds even for local development).

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: The goal is the latter--we don't actually want to create and submit an upgrade PR until it's time for that to be merged (which is when auditing and testing has completed). I was also a bit concerned with this organizational approach since it leads to lots of jumping around. Some possible options we have:
- Keep it as-is since the overall process is fairly short.
- Introduce an appendix with sections detailing specific tasks that need to be done and how, and update the process to point to these steps (this keeps the process itself linear but still results in jumping to/from the appendix, maybe even more than the current writing).
- Split this into two sections as you have it (though my concern with that is 'how to upgrade' doesn't actually have as strong of a completion indication compared to the other sections).

I slightly lean toward (1) since the other two _seem_ more complicated, but I'm not sold on any of the approaches. Do you have any thoughts?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Since this is a very short section, I'm not sure the subheadings really add much value. Do you find the current section process difficult to follow such that additional organizing might help?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: I'm not entirely sold on the specific organizational approach suggested, however I do agree that the overall flow of this section is difficult to follow and, being it's probably the most important section, needed a slight rewrite. I've attempted to reorganize it to have some sub-structure, and to have less redundancy. Is it clearer now or do you think there are additional changes that should be made?

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes.

**Comment by BenHenning** [comment]
Thanks @Rd4dev! I found your comments very helpful! Could you PTAL at my follow-ups and please let me know if you think any additional changes would be useful?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: I viewed the page using View File and it renders as below:

image
 
Am I viewing it correctly?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Thanks.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md:41]: Thanks.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Ah, thanks for the clarification!

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Yeah, makes sense. The current version by itself is easy to follow, so I'll resolve it. Thanks!

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [inline wiki/Upgrading-Target-Sdk-Guide.md]: can these too be updated as steps?

```suggestion
Work on fixing all identified problems from steps(3) and (4) (either by directly fixing the problems via code changes, or via coordination with other members of the team).
```

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: ah, maybe due to varied zoom?

https://github.com/user-attachments/assets/e74f2c87-c4d3-4ef6-b988-fe8f396cbf16

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Thanks!

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Thanks for adding the sub sections, those are helpful!

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Makes sense. I was considering whether that might create a gap in the flow..., but as you pointed out, sticking with option 1 seems like the better approach (for now).

**Review by Rd4dev — COMMENTED** [review]
Thanks @BenHenning for adding the subsections, and reg. the 'How to upgrade' section, as you mentioned it might be better to stick with the original approach for now and revisit later if needed. 

The graphs seem to render differently at various zoom levels. Please resolve any comments if they're blocking the merge.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: This is actually really interesting. What browser and browser version are you using? This might be a browser-specific GitHub and/or Mermaid bug as zoom doesn't affect things at all for me. From inspecting the graph, it's a generated SVG being rendered.

I'm using Chrome 129.

**Review by BenHenning — COMMENTED** [review]
Checked latest changes.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Good catch! Fixed.

**Review by BenHenning — COMMENTED** [review]
Thanks @Rd4dev! I followed up to both comments--PTAL.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Thanks.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: I initially tested it on Edge 128, but now I also tested it with
- Chrome 120
- Edge 128
- Chrome 129 (Updated)

seems like on my end, they were still reproducible.

https://github.com/user-attachments/assets/967ae445-7159-4abf-86b3-9039787f6964

**Comment by Rd4dev** [comment]
Thanks, @BenHenning! The only thing left is the zoom levels. I've attached a video reference as I still see the zoom level affecting them. Can you PTAL!

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Could any of the OS specific rendering make these differences, I tested it by installing Mint on VM and resembles as you mentioned but the same with Windows 11 renders it differently. Also, this issue - https://github.com/typora/typora-issues/issues/3291#issuecomment-680117090 seems to mention a similar issue for MacOs.

https://github.com/user-attachments/assets/e9c581cb-29c6-4012-822f-506cb1df96fe

# 

But Firefox renders it properly on both VM Mint OS and Windows 11 (which might suggest the earlier observations are less relevant.).

Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: I'm not 100% sure honestly. Browser rendering can get very complicated, so I can only guess on potential compatibility. I tried searching the Mermaid GitHub repo for similar issues and didn't find a perfect match (though there were other text cut-off issues for separate reasons that were addressed). The issue you linked is much closer to what we're observing here than anything I found.

I'm not opposed to adding breaks, but I have little confidence that will fix the problem correctly since a separate display situations might still result in it being cut off. Or, put another way, how much text can we actually fit on a line before it's likely to be cut off? This is further complicated by us not using a monospaced font so line width won't be exactly correlated to number of characters.

Do you have any suggestions on what we could do here? I'm wondering if a different diagram format or maybe just trying to shorten the text would be best (though I thought of that and couldn't come up with good alternatives at the time).

Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: I tried applying the `text-wrap` property to the node, and it appears to be working. Could this be a viable / consistent workaround?

Tested on few browsers and added them below, could you PTAL!

Snippet:

```
classDef textWrap text-wrap: wrap;

  A(Part 1: With Text Wrap property set to the node):::textWrap -->|Tracking issue exists| B
  B(Part 1: Without Text Wrap property set to the node) -->|Tracking issue exists| C
```

>Tested on Edge 128 

https://github.com/user-attachments/assets/6147dcd5-6824-4940-9747-aea832323d2c


<details open><summary>Chrome 129</summary>

image

</details>

<details><summary>Edge128</summary>

image

</details>

<details><summary>Firefox 130</summary>

image

</details>

<details><summary>Brave 1.70</summary>

image

</details>

**Review by adhiamboperes — APPROVED** [review]
Thanks @BenHenning! I think this guide is thorough and well structured. LGTM!

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (adhiamboperes) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Just mentioning that I have not experienced any cutoff issues on chrome or safari even when zooming.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes.

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (BenHenning) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Very nice suggestion! Applied the suggested changes. PTAL @Rd4dev.

**Review by BenHenning — COMMENTED** [review]
Thanks @Rd4dev and @adhiamboperes!

@Rd4dev I think I addressed your open comment--can you please confirm?

File: wiki/Upgrading-Target-Sdk-Guide.md
Comment (Rd4dev) [reply wiki/Upgrading-Target-Sdk-Guide.md]: Thanks, @BenHenning.

**Comment by Rd4dev** [comment]
Thanks @BenHenning for the update. I’ve resolved all my open comments.

**Comment by BenHenning** [comment]
Thanks @Rd4dev!

**Pull request merged**
